### ISSUE TITLE
Use fast 3D version of FB basis for Cov3D  

### ISSUE THREAD

**Issue by junchaoxia** [ComputationalCryoEM/ASPIRE-Python#373]
Current the CLICK command and tutorial examples for Cov3D are using the slow version of FB basis for 3D expanding. We need to switch to the fast version FFBBasis3D. However, there are some bugs to be fixed such as mismatchs of data types and Numpy array shapes as shown below: 

> File "/tigress/junchaox/software/ASPIRE-Python_git/tutorials/examples/cov3d_experiment.dontrun", line 54, in
mean_est = mean_estimator.estimate()
File "/tigress/junchaox/software/ASPIRE-Python_git/src/aspire/reconstruction/estimator.py", line 69, in estimate
est_coeff = self.conj_grad(b_coeff, tol=tol)
File "/tigress/junchaox/software/ASPIRE-Python_git/src/aspire/reconstruction/estimator.py", line 124, in conj_grad
operator, b_coeff, M=M, callback=cb, tol=tol, atol=0
File "", line 2, in cg
File "/home/junchaox/.conda/envs/aspire/lib/python3.6/site-packages/scipy/_lib/_threadsafety.py", line 46, in caller
return func(a, **kw)
File "/home/junchaox/.conda/envs/aspire/lib/python3.6/site-packages/scipy/sparse/linalg/isolve/iterative.py", line 315, in cg
work[slice2] += sclr1matvec(work[slice1])
TypeError: Cannot cast ufunc add output from dtype('complex64') to dtype('float32') with casting rule 'same_kind'

> covar_est = covar_estimator.estimate(mean_est, noise_variance)
File "/tigress/junchaox/software/ASPIRE-Python_git/src/aspire/covariance/covar.py", line 93, in estimate
b_coeff = self.src_backward(mean_vol, noise_variance)
File "/tigress/junchaox/software/ASPIRE-Python_git/src/aspire/covariance/covar.py", line 188, in src_backward
covar_b_coeff = self.basis.mat_evaluate_t(covar_b)
File "/tigress/junchaox/software/ASPIRE-Python_git/src/aspire/basis/basis.py", line 160, in mat_evaluate_t
return mdim_mat_fun_conj(X, len(self.sz), 1, self.evaluate_t)
File "/tigress/junchaox/software/ASPIRE-Python_git/src/aspire/utils/matrix.py", line 146, in mdim_mat_fun_conj
X = f(X)
File "/tigress/junchaox/software/ASPIRE-Python_git/src/aspire/basis/ffb_3d.py", line 391, in evaluate_t
v = v.reshape((*sz_roll, self.count))
ValueError: cannot reshape array of size 50176 into shape (8,8,8,8,98)

**Comment by garrettwrong** [comment]
This has two topics. First  FFB3D was apparently never used for anything outside its own unit test yet, so it naturally has some bugs.  I have some of this reconciled in `ffb3d_debugging` branch, but it's not clean enough to make a PR yet. Also...

Second issue is performance.  In some old code from April the cov3d simulation based tutorial example using FB3D took about 1 minute, and swapping to using FFB3D crashes.  Fixing it up enough to run, it took over 7 minutes.   Currently in my debugging branch, FB3D still takes about a minute, and FFB 2:20 or so.  Between not crashing out of the box and shaving off five minutes that is moving in the right direction, but still not "fast". I am trying to understand where the time goes.

![Screenshot from 2021-02-19 08-50-06](https://user-images.githubusercontent.com/47759732/108512736-a50e4e00-728f-11eb-8928-045ceb2519b0.png)

**Comment by janden** [comment]
So I'm comparing FB vs. FFB in Octave and am seeing similar timings. Not sure what's going on here. There is a speedup when you go to large images, but it's quite small. Will investigate further.

**Comment by garrettwrong** [comment]
That's good in a way I guess.

Okay, we can give back considerable time by skipping the `cb` function which duplicates expensive computations.. or call it less frequently.  I see some of my favorite confusing functions in here :D.

**Comment by janden** [comment]
Update: I originally benchmarked the het3d code, but found out that it's not as optimized as the recent Aspire code, so I update the results below.

So I put together a little benchmark script in Matlab for the Aspire code.
```
M = 1024;

basis = fb_basis(sz);
fbasis = ffb_basis(sz);

x = randn([sz M]);

tic; v = basis.evaluate_t(x); tm_evaluate_t = toc;
tic; v = fbasis.evaluate_t(x); tm_fast_evaluate_t = toc;

fprintf('evaluate_t: %f\n', tm_evaluate_t);
fprintf('evaluate_t (fast): %f\n', tm_fast_evaluate_t);

tic; xp = basis.evaluate(v); tm_evaluate = toc;
tic; xp = fbasis.evaluate(v); tm_fast_evaluate = toc;

fprintf('evaluate: %f\n', tm_evaluate);
fprintf('evaluate (fast): %f\n', tm_fast_evaluate);
```
Doing a few runs on a Flatiron workstation (dual 6-core Xeons, 3.4 GHz), it seems like breakeven is around 32:
```
>> sz = [8, 8, 8];
>> benchmark      
evaluate_t: 0.056764
evaluate_t (fast): 1.541305
evaluate: 0.048820
evaluate (fast): 2.375296
>> sz = [16, 16, 16];
>> benchmark
evaluate_t: 2.024271
evaluate_t (fast): 8.307120
evaluate: 4.131281
evaluate (fast): 4.878558
>> sz = [32, 32, 32];
>> benchmark
evaluate_t: 82.765761
evaluate_t (fast): 69.069762
evaluate: 140.859823
evaluate (fast): 36.259024
```
Although `evaluate` and `evaluate_t` have very different behaviors here.

For 2D, breakeven is around the same resolution:
```
>> sz = [16, 16];
>> benchmark     
evaluate_t: 0.029203
evaluate_t (fast): 0.080718
evaluate: 0.033603
evaluate (fast): 0.109009
>> sz = [32, 32];
>> benchmark     
evaluate_t: 0.174873
evaluate_t (fast): 0.230370
evaluate: 0.201272
evaluate (fast): 0.193703
>> sz = [64, 64];
>> benchmark     
evaluate_t: 1.302695
evaluate_t (fast): 0.787378
evaluate: 2.349181
evaluate (fast): 0.620058
```

I seem to remember these breakeven points happening at lower resolutions, but this is clearly not the case.

**Comment by garrettwrong** [comment]
Awesome (well the verification, not re the size itself), thanks. I'll try changing the problem size in the tutorial.

**Comment by janden** [comment]
> I'll try changing the problem size in the tutorial.

I don't know that that's necessary. We'd like the tutorial to run fast, so maybe smarter to switch back to the “slow” FB basis for now (I know…).

**Comment by garrettwrong** [comment]
Sure, the small tutorial is already (still) using the FB3D. I didn't mean to change it permanently, rather locally, to verify performance.  I still think this code merits some improvement. I am not able to run 32 cubed on my desktop (I only have 32GB RAM), but 16 is running now in background.  I hope the fast version is faster than this :D. It uses more memory than I expected too.

**Comment by garrettwrong** [comment]
Alright, so at (16, 16, 16) I achieve a genuine 2x speedup on my desktop.  FB approx 6 hours. FFB approx 3 hours.  Unfortunately I cannot compare results, the checkout from april crashes for any volume that is not (8,8,8).

Considering the tutorial takes two minutes using FFB3d vs one minute FB3d, then it doesn't really concern me which one we ship (from a timing perspective).

**Comment by janden** [comment]
Wait now I'm confused. Decomposing a 3D volume of size `(16, 16, 16)` takes several hours on your machine? Is this for a large stack or using a single core? Still, I don't see how it should take quite that long…

**Comment by garrettwrong** [comment]
4C/8T. The cores are pretty fast on this one...

As I read the tutorial this is 2 volumes (16, 16, 16). 1024 images (16, 16).
 
I can run on one of my rack machines or tigercpu to get a many core time.  I expect there would be a speedup switching to MKL backend for numpy as well.

But yes, this code is slower and uses a lot more memory than I was expecting.  Because the older checkout of the python code appears to just continually be broken at different depths, I think the next best comparison would require going back to matlab.

**Comment by garrettwrong** [comment]
while im waiting in princeton's queue (another reason I like my desktop :D ) I'll rerun the 16,16,16 case with a profiler.

**Comment by garrettwrong** [comment]
Still waiting for the TigerCPU node to finish (20 cores). It has already surpassed my desktop time. I might run it again with a couple slurm tweaks. I can also run it again with MKL as time permits... that all needs to be bg stuff though... I sort of need to clean up the code changes and get moving forward.. Feels like didn't make much progress today.

Anyway, the profile output from my desktop is attached. Also a quick graphic.  We are dominated by fft, ifft, and evaluate_t.  Also we spend about 1/3 the run time generating diagnostic prints in cg's `cb`. 

![FFB3d_16](https://user-images.githubusercontent.com/47759732/108759081-e6b22980-7519-11eb-8b59-d1e6027426f6.png)


[profile_ffb_16.out.zip](https://github.com/ComputationalCryoEM/ASPIRE-Python/files/6024636/profile_ffb_16.out.zip)

**Comment by janden** [comment]
All right so the running times you're quoting are for the whole tutorial at different resolutions, not just decomposing single 3D volumes. That accounts for part of the discrepancy. It might be easier to just focus on `evaluate` and `evaluate_t` for now.

**Comment by garrettwrong** [comment]
Sure, in the future I can time those methods for tuning.  To round out the end to end measurements, I ran on TigerCPU as we ship, and it was slower than my desktop, 291m.

Using MKL for numpy and fft backends on tigercpu  we get significant reduction to 173m.

**Comment by janden** [comment]
Ok thanks.

**Comment by garrettwrong** [comment]
The broken functionality was corrected with #390 .  A separate issue (#393) suggesting future performance tuning was created.

Closing.
